core/shared/src/main/scala/it/unibo/scafi/lib/Gradients.scala

```diff
@@ -5,26 +5,269 @@
+import scala.math.Numeric.DoubleIsFractional
+import it.unibo.utils.Filters.expFilter
+
 trait StdLib_Gradients {
   self: StandardLibrary.Subcomponent =>
 
   type Metric = ()=>Double
 
+  import Builtins.Bounded
+
+  implicit val idBounded: Bounded[ID]
+
   trait Gradients {
-    self: FieldCalculusSyntax with StandardSensors =>
+    self: FieldCalculusSyntax with StandardSensors with GenericUtils with StateManagement with BlockG =>
 
-    case class Gradient(algorithm: (Boolean,()=>Double)=>Double, source: Boolean=false, metric: Metric = nbrRange) {
+    case class Gradient(algorithm: (Boolean, () => Double) => Double, source: Boolean = false, metric: Metric = nbrRange) {
       def from(s: Boolean): Gradient = this.copy(source = s)
+
       def withMetric(m: Metric): Gradient = this.copy(metric = m)
+
       def run(): Double = algorithm(source, metric)
     }
 
-    val ClassicGradient = Gradient(classicGradient(_,_), false, nbrRange)
+    val ClassicGradient: Gradient = Gradient(classicGradient, source = false, nbrRange)
 
     def classicGradient(source: Boolean, metric: () => Double = nbrRange): Double =
-      rep(Double.PositiveInfinity){ case d =>
-        mux(source){ 0.0 }{ minHoodPlus(nbr(d) + metric()) }
+      rep(Double.PositiveInfinity) { case d =>
+        mux(source) {
+          0.0
+        } {
+          minHoodPlus(nbr(d) + metric())
+        }
       }
-  }
 
+    def hopGradient(source: Boolean): Double =
+      rep(Double.PositiveInfinity) {
+        hops => {
+          mux(source) {
+            0.0
+          } {
+            1 + minHood(nbr {
+              hops
+            })
+          }
+        }
+      }
+
+    def bisGradient(
+                     source: Boolean,
+                     metric: Metric = nbrRange,
+                     commRadius: Double = 0.2,
+                     lagMetric: => Double = nbrLag().toMillis): Double = {
+      val avgFireInterval = meanCounter(deltaTime().toMillis, 1000000)
+      val speed = 1.0 / avgFireInterval
+
+      rep((Double.PositiveInfinity, Double.PositiveInfinity)) { case (spatialDist: Double, tempDist: Double) =>
+        mux(source) {
+          (0.0, 0.0)
+        } {
+          minHoodPlus {
+            val newEstimate = Math.max(nbr {
+              spatialDist
+            } + metric(), speed * nbr {
+              tempDist
+            } - commRadius)
+            (newEstimate, nbr {
+              tempDist
+            } + lagMetric / 1000.0)
+          }
+        }
+      }._1
+    }
+
+    def crfGradient(
+                     source: Boolean,
+                     metric: Metric = nbrRange,
+                     raisingSpeed: Double = 5,
+                     lagMetric: => Double = nbrLag().toMillis): Double =
+      rep((Double.PositiveInfinity, 0.0)) {
+        case (g, speed) =>
+          mux(source){ (0.0, 0.0) }{
+            implicit def durationToDouble(fd: FiniteDuration): Double = fd.toMillis.toDouble / 1000.0
```

Comment:
Working with seconds seems fragile.
I see that this code came from `demos`, but here we are trying to provide a generic functionality